### PR TITLE
Added support for Intel Power Gadget

### PR DESCRIPTION
**Pull request by philipheinisch**
Added support for Intel Power Gadget. Right now only Windows (CPU&GPU) is supported. Linux support is possible but still work in progress.

### CODE REVIEW

**Comment by ranocha** [comment]
It might be good to add some CI tests for this. On Appveyor, the windows builds run on Intel CPUs, cf. https://ci.appveyor.com/project/ranocha/toolkiticl/builds/22608195/job/fuwfknym9x7dah60#L140.

**Comment by philipheinisch** [comment]
I added a very basic MSR power interface for Linux, just for testing. Do you get power values for one of the sockets or just the overall package? There might be some problem with the Intel register offsets.

**Comment by ranocha** [comment]
What do I have to do/install in order to make this work? I don't have any `/dev/cpu/[0-9]`:
```
$ ll -la /dev/cpu
total 0
drwxr-xr-x  2 root root      60 Feb 27 13:17 ./
drwxr-xr-x 21 root root    4620 Feb 27 13:17 ../
crw-------  1 root root 10, 184 Feb 27 13:17 microcode
```
Hence, I get the same error as the tests on Appveyor & Travis.

**Comment by philipheinisch** [comment]
You need the the `msr-tools` package and maybe `sudo modprobe msr`.

**Comment by ranocha** [comment]
> You need the the `msr-tools` package and `sudo modprobe msr`.

That fixed the problem for me. However, I had to run `sudo toolkitICL`, so we might want to discuss whether the default option should be to disable this and whether there are possibilities to disable errors if isp is not used.

<details>
<summary>Anyway, here is the output of some test:</summary>

```julia
julia> using HDF5

julia> h5open("out_DP_256_test.h5", "r") do io
           for n in names(io)
               println(n)
               display(read(io, n))
               println()
           end
       end
Data
Dict{String,Any} with 3 entries:
  "B" => Float32[4.0 4.0 … 0.0 0.0]
  "A" => Float32[6.0 6.0 … 0.0 0.0]
  "C" => Float32[2.4e7 0.0]

Data_LoadTime
5.479

Data_StoreTime
23.68

Global_Range
1×3 Array{Int32,2}:
 1000192  1  1

Host_OS
"Linux Ubuntu 18.04.2 LTS/4.15.0-45-generic/#48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019"

Intel_HK
Dict{String,Any} with 5 entries:
  "DRAM"       => [0.563965 0.456543 … 0.466309 0.458984]
  "Package"    => [115.625 120.781 … 95.0781 95.7813]
  "Socket 0"   => [104.766 109.844 … 84.0625 83.9258]
  "Socket 1"   => [0.0 0.0 … 0.0 0.0]
  "Power_Time" => ["27-02-2019 15:12:25:052", "27-02-2019 15:12:25:153", "27-02-201…

Kernel_ExecStart
"27-02-2019 15:12:32"

Kernel_ExecTime
7458.389

Kernel_Settings
"-DREDUCE=1 -DNR=1 -DNC=1 -DNI=1 -DREAL=float -DW_SIZE=256 -cl-single-precision-constant -cl-mad-enable -cl-no-signed-zeros -cl-finite-math-only"

Local_Range
1×3 Array{Int32,2}:
 256  1  1

NV_HK
Dict{String,Any} with 0 entries

OpenCL_Device
"Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz"

OpenCL_Version
"OpenCL 2.1 (Build 0)"

Range_Start
1×3 Array{Int32,2}:
 0  0  0

Total_ExecTime
7643.304
```
</details>
Does that answer your question?

**Comment by ranocha** [comment]
<details>
<summary>Runnign a test on a virtual machine (Ubuntu Xenial) yields</summary>

```
$ h5dump -g /Intel_HK out_DP_256_test.h5 
HDF5 "out_DP_256_test.h5" {
GROUP "/Intel_HK" {
   DATASET "DRAM" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 8, 1 ) / ( 8, 1 ) }
      DATA {
      (0,0): 0,
      (1,0): 0,
      (2,0): 0,
      (3,0): 0,
      (4,0): 0,
      (5,0): 0,
      (6,0): 0,
      (7,0): 0
      }
   }
   DATASET "Package" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 8, 1 ) / ( 8, 1 ) }
      DATA {
      (0,0): 0,
      (1,0): 0,
      (2,0): 0,
      (3,0): 0,
      (4,0): 0,
      (5,0): 0,
      (6,0): 0,
      (7,0): 0
      }
   }
   DATASET "Power_Time" {
      DATATYPE  H5T_STRING {
         STRSIZE 24;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 8 ) / ( 8 ) }
      DATA {
      (0): "27-02-2019 15:35:43:327", "27-02-2019 15:35:48:327",
      (2): "27-02-2019 15:35:53:337", "27-02-2019 15:35:58:337",
      (4): "27-02-2019 15:36:03:337", "27-02-2019 15:36:08:338",
      (6): "27-02-2019 15:36:13:338", "27-02-2019 15:36:18:338"
      }
   }
   DATASET "Socket 0" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 8, 1 ) / ( 8, 1 ) }
      DATA {
      (0,0): 0,
      (1,0): 0,
      (2,0): 0,
      (3,0): 0,
      (4,0): 0,
      (5,0): 0,
      (6,0): 0,
      (7,0): 0
      }
   }
   DATASET "Socket 1" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 8, 1 ) / ( 8, 1 ) }
      DATA {
      (0,0): 0,
      (1,0): 0,
      (2,0): 0,
      (3,0): 0,
      (4,0): 0,
      (5,0): 0,
      (6,0): 0,
      (7,0): 0
      }
   }
}
}
```
</details>

Edit: Click on the triangle to expand the output.

**Comment by philipheinisch** [comment]
It might be necessary to set the /dev/cpu permissions once, afterwards it should work.
The results seem strange for an 8700K and the DRAM consumption is very low. It is possible the DRAM power estimation is no supported on your system, but >100W is implausible for your system. I will recheck the MSR registers. Thanks.

**Comment by ranocha** [comment]
No problem; I can run new tests if you wish.

Small note: The time format `"%d-%m-%Y %H:%M:%S"` is not really international, `"%Y-%m-%d %H:%M:%S"` might be better (and is consistent, since the units become smaller from left to right).

**Comment by philipheinisch** [comment]
We can definitively change the date format. I tested it using an  i5-8400 and it seems to work. Even though it seems the power draw at idle gets underestimated, it looks ok under load. Have you tired using the `-b` switch? In this case you should see the increase in power after 4 s. Maybe the high power consumption is caused by Intel TurboBoost? Could you try  increasing the execution time and see if the power drops? It also seems to work on an Amazon ec2 instance with a E5-2676 v3 but only the package power not the sockets. To check if the correct MSRs are read, I added a function to read out the TDP, which should match the TDP given by Intel.

**Comment by philipheinisch** [comment]
It also seems to work on @Kostaszki notebooks with an i5 CPU.  Maybe at this point I should compare the MSR RAPL results against real power measurements.

**Comment by ranocha** [comment]
<details>
<summary>Using `$ sudo toolkitICL -b -isp 1000 -d 1 -c SSPRK_test.h5`, I get the following results:</summary>

```julia
julia> [intel["DRAM"]' intel["Package"]' intel["Socket 0"]' intel["Socket 1"]']
95×4 Array{Float64,2}:
 0.45282    13.0098    1.9209   0.0
 0.448181   12.2617    1.18066  0.0
 0.44635    12.168     1.09961  0.0
 0.449524   12.25      1.14063  0.0
 0.453613  137.094   126.085    0.0
 0.446838   98.6055   87.5176   0.0
 0.447449   95.0059   83.9395   0.0
 0.448059   94.9941   83.9355   0.0
 0.447754   94.9414   83.8818   0.0
 0.447205   95.0527   83.8877   0.0
 0.446655   94.9238   83.8662   0.0
 0.446716   94.9336   83.8721   0.0
 0.447754   94.9414   83.8711   0.0
 0.44696    94.9277   83.873    0.0
 0.448242   94.918    83.8457   0.0
 0.446899   94.9297   83.8721   0.0
 0.44751    94.9082   83.8477   0.0
 0.447144   94.9336   83.8643   0.0
 0.447815   95.0039   83.8535   0.0
 0.446655   94.9316   83.8633   0.0
 0.447449   94.9297   83.8662   0.0
 0.446899   94.9199   83.8574   0.0
 0.446655   94.9336   83.8701   0.0
 0.446777   94.9258   83.8643   0.0
 0.447937   94.918    83.8545   0.0
 0.44696    94.9336   83.8652   0.0
 0.447144   94.9258   83.8594   0.0
 0.447571   95.002    83.8467   0.0
 0.447632   94.9258   83.8633   0.0
 0.44696    94.9258   83.8584   0.0
 0.447388   94.9277   83.8623   0.0
 0.446838   94.9141   83.8564   0.0
 0.447571   94.9336   83.8643   0.0
 ⋮                                 
 0.446716   94.9199   83.8691   0.0
 0.447449   94.918    83.8496   0.0
 0.447876   94.9355   83.8711   0.0
 0.448181   95.002    83.8398   0.0
 0.447449   94.9316   83.8652   0.0
 0.447021   94.9297   83.8652   0.0
 0.446899   94.9121   83.8496   0.0
 0.447144   94.9277   83.8574   0.0
 0.446716   94.9199   83.8555   0.0
 0.446716   94.9316   83.873    0.0
 0.447327   94.918    83.8516   0.0
 0.447266   94.9355   83.8633   0.0
 0.446838   94.9141   83.8516   0.0
 0.447144   95.0254   83.8672   0.0
 0.446899   94.916    83.8496   0.0
 0.447449   94.9277   83.8672   0.0
 0.44751    94.9199   83.8535   0.0
 0.446655   94.9238   83.8633   0.0
 0.446838   94.9238   83.8477   0.0
 0.447021   94.9395   83.8789   0.0
 0.447327   94.916    83.8613   0.0
 0.446777   94.9121   83.8379   0.0
 0.447021   94.9258   83.8574   0.0
 0.447266   95.0293   83.875    0.0
 0.447266   94.9063   83.8418   0.0
 0.447083   94.9336   83.8652   0.0
 0.447449   94.9219   83.8594   0.0
 0.447021   94.918    83.8574   0.0
 0.449036   94.9297   83.8594   0.0
 0.453003   30.6055   19.4902   0.0
 0.447144   12.3594    1.27734  0.0
 0.449402   12.4297    1.35938  0.0
 0.446533   12.3047    1.23633  0.0
```
</details>
Clearly, the power increases after 4 seconds. The TDP is 95 (stored as a double?).

Edit: Click on the triangle to expand the output.

**Comment by philipheinisch** [comment]
That looks surprisingly good! Thanks. As it seems the >100W were indeed caused by Turbo-Boost - interesting. The TDP is also correct, so the registers get read out correctly. So apart from getting the CI checks working, the rest is mostly cleanup. How do we want to enable RAPL on Linux systems? Always enable in cmake if Linux and Intel CPUs are detected and make the user try if `-isp` works without crashing or try and detect if the required MSR files are accessible in cmake and exclude RAPL during build if not (which would mean the user needs to recompile after installing `msr-tools`)?

**Comment by ranocha** [comment]
As far as I see, there is no explicit dependency on `msr-tools` during build time. Can we restructure the code such that there is no access to msr files if the `sample_rate` for `-isp` is zero? In that case, `toolkitICL` can be used as before in most cases and the user must explicitly opt-in to log the power.

If that is possible, changes to the automatic tests are minimised. We only have to add some new tests.

**Comment by ranocha** [comment]
If the Intel and the Nvidia parts work, it might be an option to create a unified interface to the power consumption related parts, save the results in one data group and add an identifier (string), describing the method used to log the power. What do you think?

Of course, I am willing to work on that part.

**Comment by philipheinisch** [comment]
A unified power interface is a good idea. Especially if I get the AMD part working as well (work is ongoing). At some point it might even become an independent project, as a light weight cross platform power consumption interface with support for AMD, Intel and NVidia might be interesting to other people as well. Maybe we should discuss this in detail.

@ranocha do you know the best solution to get CPU temperature with Linux?

**Comment by ranocha** [comment]
> Maybe we should discuss this in detail.

That's good!

Sorry, I don't know a really good solution. On my system, 
```
$ cat /sys/class/thermal/thermal_zone2/temp
```
seems to do the trick (returns the temperature in milli °C, cf. https://www.kernel.org/doc/Documentation/thermal/sysfs-api.txt). However, there is no such file on my VMs.

**Comment by ranocha** [comment]
Did you try something like https://github.com/torvalds/linux/blob/7d762d69145a54d169f58e56d6dac57a5508debc/tools/power/x86/turbostat/turbostat.c#L1937?

**Comment by philipheinisch** [comment]
The question is, if we want to use something vendor specific like MSRs or something more generic like the sysfs drivers? A solution based on the MSRs might be the most powerful and requires very little extra dependencies. The sysfs approach is more generic and cross platform but might not work out of the box on all systems (especially clusters).

**Comment by ranocha** [comment]
Well, maybe there is simply no "best" solution. If we want to have something working on as many platforms as possible, we might be forced to do something like
```
if (sysfs_works()) {
  do_sysfs();
}
else if (msr_works()) {
  do_msr();
}
else if (...) 
```

**Comment by philipheinisch** [comment]
That might very well be the case. For now, the MSR approach might be the easiest, as the necessary MSR read functions are already implemented.

**Comment by ranocha** [comment]
If we unify the interface for power/temperature/etc. logging, we should also unify the command line options. This would also simplify the testing and application of `toolkitICL`.

**Comment by ranocha** [comment]
Adding really good CI tests might be a bit problematic, since the low-level MSR tools are not really supported in VMs. On my machines, the output is simply zero for all variables. Thus, we cannot test the output reliably. Nevertheless, I will add some tests of the basic functionality later.

**Comment by philipheinisch** [comment]
> If we unify the interface for power/temperature/etc. logging, we should also unify the command line options. This would also simplify the testing and application of `toolkitICL`.

If we unify everything, it becomes impossible to log CPU and GPU data concurrently and we would lose functionality. Especially as it might be interesting to look at the CPU power consumption when using a GPU, as the CPU still has to do maintenance and control functions during execution. Do we really want to loose that functionality?

**Comment by ranocha** [comment]
That's a good question. From the perspective of a user, having a single command line option determining the power logging for all devices is nice, because only one parameter has to be changed if another device should be used.

I see your point from from the developer perspective. A reasonable compromise might be to have both a general option and specific logging options for each device/manufacturer/method of measurement (including command line options and output groups/directories).

In the end, we could have something working as follows:
- `toolkitICL -sp 1000 -d device -c config.h5`
  Chooses the backend based on the OpenCL `device` and writes the result to the group `log`.
- `toolkitICL -isp 1000 -d device -c config.h5`
  Logs the CPU power consumption using the Intel MSR tools and writes the result to the group `intel_log`. Similar for NVidia, AMD, sysfs, ...
- `toolkitICL -isp 1000 -sp 1000 -d device -c config.h5`
  Chooses a backend based on the OpenCL `device` to write results to the group `log` and logs the CPU power consumption using the Intel MSR tools, writing the result to the group `intel_log`.

The unified interface could also be added later. It should determine the specific implementation based on the chosen OpenCL device somehow.

In order to simplify some scripts right now, we could detect how many `-*sp` command line options are given. If only one logging command is present, the results can be written to the group `log`. Otherwise, the results should be written to a group `intel_log`, `nvidia_log` etc.

Of course, the naming scheme for the options/groups is still open to discussions.

**Comment by philipheinisch** [comment]
While it would be very user friendly, choosing the implementation automatically based on the device is not straight forward. One would not only need to check OS but also the the device vendor and type. For example with integrated graphics, the driver vendor and version can be the same for CPU and GPU. It could also cause problems with POCL or other 3rd party implementations, as it sometimes just reports back as `pthread xxx`. The Intel OpenCL CPU drivers also work (very well) with AMD CPUs but reports `Intel` as driver vendor, so one would need to parse the device name and match it with known Intel or AMD devices. AMD drivers also work with Intel CPUs albeit with bad performance. Most ARM processors also provide enhanced power information, but it would make this approach even more complicated. This would be a brilliant feature for an easy-to-use power library, but I think it would not be time effective to implement this right now as part of toolkitICL. Maybe we could choose the command line switches, similar to what you suggested, in such a way, that it is future proof and we can add an automatic `-sp` switch later and right now just use device specific flags?

**Comment by ranocha** [comment]
That's okay for me.

**Comment by ranocha** [comment]
I will add some CI tests today (later). Because of

> While it would be very user friendly, choosing the implementation automatically based on the device is not straight forward. One would not only need to check OS but also the the device vendor and type.

CMake tests (which would have to implement the functionality mentioned there) do net seem useful now. However, adding some tests to the CI scripts directly seems possible.

**Comment by ranocha** [comment]
This seems to need some debugging: Using the repetition test, I get the following result on Travis (https://travis-ci.org/ranocha/toolkitICL/builds/501609313):
```
$ cd test; sudo ../bin/toolkitICL -b -isp 1000 -c repetition_test.h5

Benchmark mode
Available devices: 1
                 Intel(R) Xeon(R) CPU @ 2.50GHz
OpenCL version: OpenCL 1.2 (Build 37)
Memory limit: 7835996160
WG limit: 8192
Reading kernel from file: add_one_kernel.cl... 
Found Kernels: 1
Number of Kernels to execute: 5
Ingesting HDF5 config file...
Creating output HDF5 file...
Old HDF5 data file found and deleted!
Setting range...
read_msr():pread: Input/output error
```

**Comment by philipheinisch** [comment]
Could be a problem with virtualization. I get similar problems using VirtualBox. The error message seems like a low-level msr-tools error.

**Comment by ranocha** [comment]
Hm, strange. I have no problems using VirtualBox.

**Comment by ranocha** [comment]
It seems to be better on Appveyor (https://ci.appveyor.com/project/ranocha/toolkiticl-3vn05). However, `Intel_HK/GT0` is not present in the output file. Should this info be required or can I drop it?

Edit: Okay, I've found the relevant part suggesting that it should not be required.

**Comment by philipheinisch** [comment]
GT0 is only present, if an internal Intel GPU is detected. So it's probably fine.

**Comment by ranocha** [comment]
Okay. Hopefully, the tests pass also for this repository. Let's see.

**Comment by ranocha** [comment]
The tests pass. These are some simple build & interface tests. Testing whether the returned values are meaningful is not possible on the virtual machines, cf. https://ci.appveyor.com/project/ranocha/toolkiticl/builds/22812170/job/nt8fxlwebsxg0xmj#L2190.

**Comment by ranocha** [comment]
Results of our internal discussion:
- [x] Unify the units (milliwatt, milliseconds)
- [ ] AMD is WIP
- [x] Command line options: `-intel_power sample_rate`, `-nvidia_power sample_rate`, `-nvidia_temperature sample_rate`, `-amd_power sample_rate`?
- [ ] Choose an Intel implementation for Mac?
- [x] Output names: `Housekeeping/intel`, `Housekeeping/nvidia`
- [x] Output `Kernel_ExecStart`: add milliseconds
- [ ] Write a new `IntelPowerGadgetLib` using only the absolutely necessary parts and without source code provided by Intel
- [ ] Some cleanup (indentation, tabs/spaces...)

**Comment by ranocha** [comment]
It might make sense to store date/time values as strings using ISO 8601 in this PR unless there is a workaround for the problems mentioned in #21. That would be `YYYY-MM-DDThh:mm:ss.sss`, e.g. `2019-03-06T18:25:40.123`. This would result in human readable time values.

**Comment by ranocha** [comment]
If we want to use binary representations of dateimes, the best possibility I have found after some search may be to use the POSIX seconds as `Float64`/`double`. This can be explained in an attribute added to datetime values (#23).

Here are some examples using high level languages.
- `python`
  Documentation: https://docs.scipy.org/doc/numpy/reference/arrays.datetime.html
  ```python
  In [1]: import numpy as np

  In [2]: dt = np.datetime64('2019-03-06T18:25:40.123')

  In [3]: dt.tobytes()
  Out[3]: '\x1b\x01ATi\x01\x00\x00'

  In [4]: dt.dtype
  Out[4]: dtype('<M8[ms]')

  In [5]: dt.astype('float64') / 1e3
  Out[5]: 1551896740.123

  In [6]: np.float64(1551896740.123 * 1e3).astype('<M8[ms]')
  Out[6]: numpy.datetime64('2019-03-06T18:25:40.123')
  ```
  Further Reading: https://github.com/pandas-dev/pandas/pull/11628

- `julia`
  Documentation: https://docs.julialang.org/en/v1/stdlib/Dates/index.html
  ```julia
  julia> using Dates

  julia> dt = DateTime("2019-03-06T18:25:40.123")
  2019-03-06T18:25:40.123

  julia> dt.instant.periods.value
  63687579940123

  julia> reinterpret(UInt8, [dt.instant.periods.value])
  8-element reinterpret(UInt8, ::Array{Int64,1}):
  0x1b
  0x35
  0x94
  0x6b
  0xec
  0x39
  0x00
  0x00

  julia> datetime2unix(dt)
  1.551896740123e9

  julia> unix2datetime(1.551896740123e9)
  2019-03-06T18:25:40.123
  ```

- `matlab`
  Documentation: https://de.mathworks.com/help/matlab/ref/datetime.html and https://de.mathworks.com/help/matlab/ref/datetime.posixtime.html
  ```matlab
  >> dt = datetime('2019-03-06T18:25:40.123', 'InputFormat', 'yyyy-MM-dd''T''HH:mm:ss.SSS')

  dt =

    datetime

    06-Mar-2019 18:25:40

  >> format long
  >> posixtime(dt)

  ans =

      1.551896740123000e+09

  >> datetime(1.551896740123000e+09, 'ConvertFrom', 'posixtime')

  ans =

    datetime

    06-Mar-2019 18:25:40
  ```

The direct binary representations of `julia` and `python` do not seem to match and I don't know how to get a useful binary representation in `matlab`. Hence, I vote for POSIX time as `Float64`/`double`.

**Comment by philipheinisch** [comment]
So do we want to use human readable strings, or POSIX date format? Aren't date strings easier to read in general? Maybe we separate this into an independent PR?

**Comment by ranocha** [comment]
That's a good idea. I'll open an issue so that we can discuss which format we want to have. In this PR, adding milliseconds to `Kernel_ExecStart` suffices.

However, using the ISO 8601 format `2019-03-06T18:25:40.123` instead of `2019-03-06 18:25:40:123` might nevertheless be a good idea and could be added in this PR.

**Comment by philipheinisch** [comment]
We should use the ISO 8601 format in any case. I will add the milliseconds. Anyway we should take care to minimize merge conflicts between PRs.

File: src/main.cpp [hunk 1]
```diff
@@ -547,6 +791,124 @@ int main(int argc, char *argv[]) {
...
+#if defined(USEIRAPL)
+
+	 is_log_pwr = false;
+	 is_log_pwr_thread.join();
+
+ if (is_p_rate > 0)
+ {
+	 std::vector<std::string> time_strings;
+
+	 for (size_t i = 0; i < is_pwr_time.size()-1; i++) {
```
Comment (ranocha) [inline src/main.cpp]: Why do you skip the last element here?

Edit: I see, there are some differences between consecutive elements below....

Comment (philipheinisch) [reply src/main.cpp]: Because I start counting at zero? And the range of the vecotr is [0..is_pwr_time.size()-1]. Or did I miss something?

Comment (ranocha) [reply src/main.cpp]: Stating at zero is okay if you use `for (size_t i = 0; i < is_pwr_time.size(); i++) {`. The reason seems to be that there are differences `is_pwr0[i].at(j+1)-is_pwr0[i].at(j)` below.

Comment (philipheinisch) [reply src/main.cpp]: Ah of course, as the logged values are energy and not power, I have to take the difference (or derivative) to get the power. Hence, I also had to subtract one from the time to get the same length vector. I did not want to calculate the difference directly in the logging thread to save CPU time.

Comment (philipheinisch) [reply src/main.cpp]: The timestamps are logged accordingly in the thread to account for this circumstance.

Comment (ranocha) [reply src/main.cpp]: Isn't energy the more interesting quantity in the end?

Comment (philipheinisch) [reply src/main.cpp]: The problem is that the implementations that measure power directly and do not calculate it heuristically output power as direct measurement and not energy. Therefore, we can not easily compare the energy values. The energy counter is also not save against overflow. In the final implementation, we should probably give the users access to all available values.

Comment (ranocha) [reply src/main.cpp]: Okay.

**Comment by ranocha** [comment]
The tests seem to be good now. I'll copy the remaining items from the TODO list above https://github.com/IANW-Projects/toolkitICL/pull/18#issuecomment-469826865:

- [x] Unify the units (power in watt as `Float32`, time in seconds as `Float64`, temperature in °C as `int16`)
- [ ] AMD is WIP; proably `-amd-cpu_` and `-amp_gpu`?
  - [x] Unify the units
  - [x] Rename internal logging variables
- [ ] Choose an Intel implementation for Mac? -> Later, cf. #29.
- [ ] Write a new `IntelPowerGadgetLib` using only the absolutely necessary parts and without source code provided by Intel
- [x] Rename internal logging variables according to the new command line options?
- [ ] Some cleanup (indentation, tabs/spaces...)

I can unify the units, rename internal logging variables, and do some cleanup. But we should discuss which units we want to have before.

**Comment by ranocha** [comment]
I didn't touch the AMD part because @philipheinisch is working on that.

**Comment by philipheinisch** [comment]
Unfortunately, I now have to merge by hand, as I was also changing parts of the code you changed. Could you maybe wait until I am finished, as this is not time critical?

**Comment by ranocha** [comment]
Okay, sorry.

**Comment by philipheinisch** [comment]
The AMD CPU interface now works using both Windows and Ubuntu, AMD GPU support is still WIP. Both the AMD and Nvidia implementations have been tested.

**Comment by ranocha** [comment]
Naming schemes of the output file:
- All lower case and underscores
- Directory `architecture`: Host_OS, OpenCL_Device, OpenCL_Platform, OpenCL_Version
- `settings`: Global_Range, Local_Range, Range_Start, Kernel_Settings
- `housekeeping`: additionally Data_LoadTime, Data_StoreTime, Kernel_ExecStart, Kernel_ExecTime, Total_ExecTime
- `execution_time` etc.

Input file analogously.

Cf. https://github.com/philipheinisch/toolkitICL/pull/3#issuecomment-471277324

**Comment by ranocha** [comment]
Add `TESTS` option to cmake (`ccmake`), cf. #28.

**Comment by ranocha** [comment]
Let's see whether github recognises this message: Commits included here close #21.

**Comment by philipheinisch** [comment]
Theoretically this should be enough to merge this PR or is there still something missing?

**Comment by ranocha** [comment]
I will have a look at the new commits soon.
What's the problem with concurrent power and temperature logging you mentioned earlier (in #20)?

File: src/main.cpp [hunk 2]
```diff
@@ -942,7 +1020,7 @@ int main(int argc, char *argv[]) {
 
   if (intel_log_power)
   {
-    h5_write_single<float>(out_name, "/housekeeping/intel/TDP", rapl->get_TDP(),
+    h5_write_single<uint32_t>(out_name, "/housekeeping/intel/TDP", rapl->get_TDP(),
```
Comment (ranocha) [inline src/main.cpp]: Other power data are stored as `float`. Should we really use two different data types here?

File: src/main.cpp [hunk 3]
```diff
@@ -1007,9 +1089,23 @@ int main(int argc, char *argv[]) {
...
-      h5_write_buffer<cl_float>(out_name, varname.c_str(), amd_power0[i].data(), amd_power0[i].size(),
+      //cout<<"Power: "<< AMDP_names.at(i) <<endl;
+      h5_write_buffer<cl_float>(out_name, varname.c_str(), amd_power[i].data(), amd_power[i].size(),
       "Power in watt");
     }
+
+  }
+  if (amd_temp_rate > 0)
+  {
+      h5_write_buffer<double>(out_name, "/housekeeping/amd/temperature_time", amd_power_time.data(), amd_power_time.size(),
```
Comment (ranocha) [inline src/main.cpp:1106]: Shouldn't this be `amd_temp_time` instead of `amd_power_time`?

File: src/main.cpp (hunk 2 above)
Comment (philipheinisch) [reply src/main.cpp]: No, float is ok, the drivers just return ints, so it was just the intuitive habit to change everything to the same datatype without thinking...

**Comment by ranocha** [comment]
Which encoding do you use for the `README.md`? On my system using UTF-8 as standard encoding, the `µ` in AMD µProf is displayed as `�`.

File: src/main.cpp (hunk 2 above)
Comment (ranocha) [reply src/main.cpp]: Okay. Will you change it back to `float` in that case?

File: src/main.cpp (hunk 3 above)
Comment (philipheinisch) [reply src/main.cpp:1106]: It should, if two separate threads with different sampling rates are used, but unfortunately, this does not work. So right now it is more resource efficient to use one thread with one time vector. The problem is that if both temperature and power logging are enabled, the driver only returns power. Separately it works.

**Comment by ranocha** [comment]
Do you plan to add support for power/temperature logging on AMD GPUs in this PR or in another one in the future? If it shall be implemented in the future, we should add that part to #20.

File: src/main.cpp (hunk 3 above)
Comment (ranocha) [reply src/main.cpp:1106]: Okay. Could you please add a corresponding remark in the source code and file an issue?

**Comment by philipheinisch** [comment]
Right now, it should output GPU power/temperature as well, if CPU logging is enabled, as there is no device specific selection. I just have not managed to test it yet due to driver problems. As soon as I get the drivers running, I can test how the GPU is enumerated and add a specific switch. Maybe we should amend #20?

**Comment by philipheinisch** [comment]
> Which encoding do you use for the `README.md`? On my system using UTF-8 as standard encoding, the `µ` in AMD µProf is displayed as `�`.

It seems the encoding is changed during the push process by git on my system - could you fix it?

**Comment by ranocha** [comment]
> Maybe we should amend #20?

That's a good idea.

**Comment by ranocha** [comment]
If the remark described in https://github.com/IANW-Projects/toolkitICL/pull/18#discussion_r264738035 is added and the change to `float` described in https://github.com/IANW-Projects/toolkitICL/pull/18#discussion_r264735327 is done, we can merge this PR.

Edit: Should also mention the problem with concurrent power/temperature logging for AMD in the README or the displayed help?

**Comment by philipheinisch** [comment]
Datatype is fixed and I added a warning, if concurrent AMD logging is selected.

**Review by ranocha — APPROVED** [review]

**Pull request merged**